From 703c73dc55eea208118b316cbdcdf5ba5c6549a2 Mon Sep 17 00:00:00 2001
From: CIQ Kernel Automation <ciq_kernel_automation@ciq.com>
Date: Mon, 13 Jul 2026 07:37:19 +0000
Subject: [PATCH] net: bridge: mst: fix suspicious rcu usage in
 br_mst_set_state

jira VULN-5460
cve CVE-2024-40920
commit-author Nikolay Aleksandrov <razor@blackwall.org>
commit 546ceb1dfdac866648ec959cbc71d9525bd73462

I converted br_mst_set_state to RCU to avoid a vlan use-after-free
but forgot to change the vlan group dereference helper. Switch to vlan
group RCU deref helper to fix the suspicious rcu usage warning.

Fixes: 3a7c1661ae13 ("net: bridge: mst: fix vlan use-after-free")
	Reported-by: syzbot+9bbe2de1bc9d470eb5fe@syzkaller.appspotmail.com
Closes: https://syzkaller.appspot.com/bug?extid=9bbe2de1bc9d470eb5fe
	Signed-off-by: Nikolay Aleksandrov <razor@blackwall.org>
Link: https://lore.kernel.org/r/20240609103654.914987-3-razor@blackwall.org
	Signed-off-by: Jakub Kicinski <kuba@kernel.org>
(cherry picked from commit 546ceb1dfdac866648ec959cbc71d9525bd73462)
	Signed-off-by: CIQ Kernel Automation <ciq_kernel_automation@ciq.com>
---
 net/bridge/br_mst.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/net/bridge/br_mst.c b/net/bridge/br_mst.c
index 3c66141d34d6..2e46ec1e98d3 100644
--- a/net/bridge/br_mst.c
+++ b/net/bridge/br_mst.c
@@ -103,7 +103,7 @@ int br_mst_set_state(struct net_bridge_port *p, u16 msti, u8 state,
 	int err = 0;
 
 	rcu_read_lock();
-	vg = nbp_vlan_group(p);
+	vg = nbp_vlan_group_rcu(p);
 	if (!vg)
 		goto out;
 
-- 
2.52.0

